File: AN_3/SEM_1/PPD/lab1/tema1java/src/main/java/lab/HorizontalAllocatedThreads.java
package lab;

// Clasa pentru paralelizare pe orizontala (mai multe linii / thread)
public class HorizontalAllocatedThreads {
    int N, M, n, p;
    int[][] matrix, conv, result;

    public HorizontalAllocatedThreads(int N, int M, int n, int p, int[][] matrix, int[][] conv) {
        this.N = N;
        this.M = M;
        this.n = n;
        this.p = p;
        this.matrix = matrix;
        this.conv = conv;
    }

    // Functie pentru rulare cu threaduri
    public int[][] run() throws InterruptedException {
        result = new int[N][M];
        Thread[] threads = new Thread[p];
        int rowsPerThread = N / p;       // impartirea liniilor
        int extra = N % p;               // linii ramase
        int start = 0;

        // cream threaduri
        for (int i = 0; i < p; i++) {
            int end = start + rowsPerThread + (extra-- > 0 ? 1 : 0);
            threads[i] = new Worker(start, end);
            threads[i].start();
            start = end;
        }

        // asteptam threadurile sa termine
        for (Thread t : threads) t.join();
        return result;
    }

    // Thread pentru un bloc de linii
    class Worker extends Thread {
        int startRow, endRow;
        public Worker(int startRow, int endRow) {
            this.startRow = startRow;
            this.endRow = endRow;
        }

        public void run() {
            int border = n / 2;
            for (int i = startRow + border; i < endRow + border; i++) {
                for (int j = border; j < M + border; j++) {
                    int sum = 0;
                    for (int x = 0; x < n; x++)
                        for (int y = 0; y < n; y++)
                            sum += matrix[i - border + x][j - border + y] * conv[x][y];
                    result[i - border][j - border] = sum;
                }
            }
        }
    }
}

File: AN_3/SEM_1/PPD/lab1/tema1cppdinamic/HorizontalAllocatedThreads.cpp
#include "HorizontalAllocatedThreads.h"
#include <thread>
#include <vector>
using namespace std;

HorizontalAllocatedThreads::HorizontalAllocatedThreads(int N, int M, int n, int p, int** bordered, int (*conv)[6]) {
    this->N = N;
    this->M = M;
    this->n = n;
    this->p = p;
    this->bordered = bordered;
    this->conv = conv;
}

// Functie worker care proceseaza un interval de linii
void HorizontalAllocatedThreads::worker(int startRow, int endRow, int** result) {
    int border = n / 2;

    // Parcurgem liniile alocate acestui thread
    for (int i = startRow; i < endRow; i++) {
        for (int j = 0; j < M; j++) {
            int sum = 0;

            // Aplicam convolutia pentru fiecare element
            for (int x = 0; x < n; x++)
                for (int y = 0; y < n; y++)
                    sum += bordered[i + x][j + y] * conv[x][y];
            result[i][j] = sum;
        }
    }
}

// Functie care porneste thread-urile si le aloca pe linii
void HorizontalAllocatedThreads::run(int** result) {
    vector<thread> threads;
    int rowsPerThread = N / p;
    int start = 0;

    // Cream thread-urile si le atribuim intervale de linii
    for (int i = 0; i < p; i++) {
        int end = (i == p - 1) ? N : start + rowsPerThread;
        threads.emplace_back(&HorizontalAllocatedThreads::worker, this, start, end, result);
        start = end;
    }

    // Asteptam ca toate thread-urile sa termine
    for (auto &t : threads)
        t.join();
}
